src/Http/Http/src/Timeouts/RequestTimeoutsMiddleware.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Http.Timeouts;
+
+internal sealed class RequestTimeoutsMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly IOptionsMonitor<RequestTimeoutOptions> _options;
+
+    public RequestTimeoutsMiddleware(RequestDelegate next, IOptionsMonitor<RequestTimeoutOptions> options)
+    {
+        _next = next;
+        _options = options;
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        var endpoint = context.GetEndpoint();
+        var timeoutMetadata = endpoint?.Metadata.GetRequiredMetadata<RequestTimeoutAttribute>();
+        var policyMetadata = endpoint?.Metadata.GetRequiredMetadata<RequestTimeoutPolicy>();
+
+        var options = _options.CurrentValue;
+
+        if (timeoutMetadata is null && policyMetadata?.Timeout is null && options.DefaultTimeout is null)
+        {
+            return _next(context);
+        }
+
+        var disableMetadata = endpoint?.Metadata.GetRequiredMetadata<DisableRequestTimeoutAttribute>();
+        if (disableMetadata is not null)
+        {
+            return _next(context);
+        }
+
+        TimeSpan? timespan;
+        if (policyMetadata is not null)
+        {
+            timespan = policyMetadata.Timeout;
+        }
+        else if (timeoutMetadata is not null)
+        {
+            if (timeoutMetadata.Timeout is not null)
+            {
+                timespan = timeoutMetadata.Timeout.Value;
+            }
+            else
+            {
+                if (options.Policies.TryGetValue(timeoutMetadata.PolicyName!, out var policy))
+                {
+                    timespan = policy.Timeout;
+                }
+                else
+                {
+                    // policy does not exists. what to do now!?
+                    throw new Exception();
+                }
```

Comment:
Can we resolve this in the metadata call and throw from there?

Reply:
Thats why OnChange exists.

Reply:
That's less helpful here OnChange is called from background threads where the exception would be hard to notice, as opposed to the main startup thread where it would crash the app.